lib/Sema/CSDiag.cpp

```diff
@@ -6223,11 +6223,47 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
   if (diagnoseCallContextualConversionErrors(callExpr, CS->getContextualType()))
     return true;
 
-  // Type check the function subexpression to resolve a type for it if
-  // possible.
-  auto fnExpr = typeCheckChildIndependently(callExpr->getFn());
-  if (!fnExpr)
+  auto originalFnType = CS->getType(callExpr->getFn());
+
+  // Check if there is a structural problem in the function expression
+  // itself, by type-checking it allowing unresolved variables. If
+  // that is going to produce a functional type with unresolved result
+  // let's not re-typecheck the function expression, because it might
+  // produce unrelated diagnostics due to lack of contextual information.
+  auto shouldTypeCheckFunction = [&](Expr *fnExpr) -> bool {
+    if (!isa<UnresolvedDotExpr>(fnExpr))
+      return true;
+
+    SmallVector<Type, 4> fnTypes;
+    CS->TC.getPossibleTypesOfExpressionWithoutApplying(
+        fnExpr, CS->DC, fnTypes, FreeTypeVariableBinding::UnresolvedType);
+
+    // Might be a structural problem related to the member itself.
+    if (fnTypes.empty())
```

Comment:
Do you really want to check for `empty` explicitly? Not doing so results in returning true anyway.